x/stake/handler.go

```diff
@@ -217,81 +178,78 @@ func handleMsgUnbond(ctx sdk.Context, msg MsgUnbond, k Keeper) sdk.Result {
 		}
 	} else {
 		var err sdk.Error
-		shares, err = sdk.NewRatFromDecimal(msg.Shares)
+		delShares, err = sdk.NewRatFromDecimal(msg.Shares)
 		if err != nil {
 			return err.Result()
 		}
-		if bond.Shares.LT(shares) {
+		if bond.Shares.LT(delShares) {
 			return ErrNotEnoughBondShares(k.codespace, msg.Shares).Result()
 		}
 	}
 
-	// get candidate
-	candidate, found := k.GetCandidate(ctx, msg.CandidateAddr)
+	// get validator
+	validator, found := k.GetValidator(ctx, msg.ValidatorAddr)
 	if !found {
-		return ErrNoCandidateForAddress(k.codespace).Result()
+		return ErrNoValidatorForAddress(k.codespace).Result()
 	}
 
 	if ctx.IsCheckTx() {
-		return sdk.Result{
-			GasUsed: GasUnbond,
-		}
+		return sdk.Result{}
 	}
 
 	// retrieve the amount of bonds to remove (TODO remove redundancy already serialized)
 	if msg.Shares == "MAX" {
-		shares = bond.Shares
+		delShares = bond.Shares
 	}
 
 	// subtract bond tokens from delegator bond
-	bond.Shares = bond.Shares.Sub(shares)
+	bond.Shares = bond.Shares.Sub(delShares)
 
 	// remove the bond
 	revokeCandidacy := false
 	if bond.Shares.IsZero() {
 
-		// if the bond is the owner of the candidate then
+		// if the bond is the owner of the validator then
 		// trigger a revoke candidacy
-		if bytes.Equal(bond.DelegatorAddr, candidate.Address) &&
-			candidate.Status != Revoked {
+		if bytes.Equal(bond.DelegatorAddr, validator.Owner) &&
+			validator.Revoked == false {
 			revokeCandidacy = true
 		}
 
-		k.removeDelegatorBond(ctx, bond)
+		k.removeDelegation(ctx, bond)
 	} else {
 		// Update bond height
 		bond.Height = ctx.BlockHeight()
-		k.setDelegatorBond(ctx, bond)
+		k.setDelegation(ctx, bond)
 	}
 
 	// Add the coins
-	p := k.GetPool(ctx)
-	p, candidate, returnAmount := p.candidateRemoveShares(candidate, shares)
+	pool := k.GetPool(ctx)
+	validator, pool, returnAmount := validator.removeDelShares(pool, delShares)
 	returnCoins := sdk.Coins{{k.GetParams(ctx).BondDenom, returnAmount}}
 	k.coinKeeper.AddCoins(ctx, bond.DelegatorAddr, returnCoins)
 
 	/////////////////////////////////////
 
-	// revoke candidate if necessary
+	// revoke validator if necessary
 	if revokeCandidacy {
 
 		// change the share types to unbonded if they were not already
-		if candidate.Status == Bonded {
-			p, candidate = p.bondedToUnbondedPool(candidate)
+		if validator.Status() == sdk.Bonded {
+			validator, pool = validator.UpdateStatus(pool, sdk.Unbonded)
 		}
 
-		// lastly update the status
-		candidate.Status = Revoked
+		validator.Revoked = true
 	}
 
-	// deduct shares from the candidate
-	if candidate.Liabilities.IsZero() {
-		k.removeCandidate(ctx, candidate.Address)
+	// deduct shares from the validator
+	if validator.DelegatorShares.IsZero() {
```

Comment:
Where in this function do we update the validator set accordingly (since someone else may now be a validator) when `DelegatorShares` becomes zero? Maybe I missed it, but I don't see that anywhere, and I also don't think it's covered by a testcase.

Reply:
yeah, you're becoming a staking master. It's not, moved `k.UpdateValidator` to before this call to update the validator set under a zero delegator share situation.

Reply:
okay addressed - also cleanup up a bit of the revoke logic